libyul/optimiser/UnusedStoreEliminator.h

```diff
@@ -0,0 +1,117 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Optimiser component that removes stores to memory and storage slots that are not used
+ * or overwritten later on.
+ */
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/UnusedStoreBase.h>
+
+#include <map>
+#include <vector>
+
+namespace solidity::yul
+{
+struct Dialect;
+struct AssignedValue;
+
+/**
+ * Optimizer component that removes sstore statements if they
+ * are overwritten in all code paths or never read from.
+ *
+ * The m_store member of UnusedStoreBase is only used with the empty yul string
+ * as key in the first dimension.
+ *
+ * Best run in SSA form.
+ *
+ * Prerequisite: Disambiguator, ForLoopInitRewriter.
```

Comment:
```suggestion
 * Prerequisite: Disambiguator, ForLoopInitRewriter, FunctionHoister.
```
?
Otherwise I'd need to wonder if we need to save stuff like ``m_forLoopInfo`` and/or ``m_forLoopNestingDepth`` in the function definition visit...

Reply:
Oh, sorry, no, this is actually done in ``UnusedStoreBase::operator()``.... but isn't that missing the ``m_forLoopNestingDepth`` part there?

Reply:
Ah that is true, but non-critical, it's just a resource limiter.

Reply:
Yeah, and in practice it'll be hoisted anyways, so doesn't matter.

Reply:
https://github.com/ethereum/solidity/pull/12728 if we still want to fix it.